src/scalar_store.rs

```diff
@@ -251,22 +414,568 @@ pub enum ScalarContinuation<F: PrimeField> {
     Terminal,
 }
 
+impl<F: LurkField> IpldEmbed for ScalarContinuation<F> {
+    fn to_ipld(&self) -> Ipld {
+        match self {
+            Self::Outermost => Ipld::List([Ipld::Integer(ContTag::Outermost as i128)].into()),
+            Self::Call {
+                unevaled_arg,
+                saved_env,
+                continuation,
+            } => Ipld::List(vec![
+                Ipld::Integer(ContTag::Call as i128),
+                unevaled_arg.to_ipld(),
+                saved_env.to_ipld(),
+                continuation.to_ipld(),
```

Comment:
It seems important that this remain synchronized with `get_hash_components()`. The unfortunate manual nature of these layouts is already a source of bugs in the circuit — so even if not strictly required, it will be very confusing if these drift (or do not agree initially). It also seems likely that we will directly depend on consistency at some point, when we want to generate the continuation hashes directly from these components. In fact, we should probably have a consistency check that allows us to do that now (maybe in `from_ipld`, optionally).

Minimally, let's add a comment here and in `get_hash_components` noting that these need to be kept in sync. Better would be a test enforcing that this is the case. Maybe add a TODO comment if we don't implement that now.

Reply:
So I agree that the order of arguments should remain in sync with `get_hash_components`, but I think in this case it's actually fine for there to be different ordering in the ipld as long as `to_ipld` and `from_ipld` are consistent. I will add these comments though.

One thing that we could think about to help consistency is switching from `Foo { x, y, z }` struct representation (which is technically unordered) to `Foo(x,y,z)` tuple representation. In principle this would force everyone to think about arg order, but there's probably an ergonomic tradeoff here too

Reply:
Yes, there are two independent but related issues:

1. We should enforce the same ordering as a codebase-level design decision.

2. We should ensure test coverage ensure roundtrip consistency for every type.

I think in the end, there will be a structure which allows us to guarantee correctness in an ergonomic way. This will need to provide some affordances used by the store, by circuits, and by the IPLD embedding. Manually enforcing consistency even unneeded for 'correctness' is at least a step in that direction (plus will help prevent coding errors and confusion along the way).

I don't think we're disagreeing. I'm just elaborating further.